Training24/test.c

```diff
@@ -0,0 +1,170 @@
+       "00010000",  // 65536
+       "7FFFFFFF",  // 2147483647
+       "80000001",  // -2147483647
+       "FFFFFF01",  // -255
+       "FFFFFFFF"   // -1
+   };
+   int numTests = sizeof (inputs) / sizeof (inputs[0]);
+   char** actual_hex_result = malloc (numTests * sizeof (char*));
+   int* pass = malloc (numTests * sizeof (int));
+   if (actual_hex_result == NULL || pass == NULL) {
+      printf ("Memory allocation failed\n");
+      return;
+   }
+   for (int i = 0; i < numTests; i++) {
+      char* result = Dec2Hex (inputs[i]);
+      pass[i] = (result && strcmp (result, expected_hex_result[i]) == 0) ? 1 : 0;
+      actual_hex_result[i] = result;
+   }
+   // Call the function to format and print the results
+   FormatHex (numTests, inputs, expected_hex_result, actual_hex_result, pass);
+   for (int i = 0; i < numTests; i++) free (actual_hex_result[i]);
+   free (actual_hex_result);
+   free (pass);
+}
+
+int main () {
+   // Run test cases
+   TestDec2Bin ();                                                          // Call test function for decimal to binary conversion      
+   TestDec2Hex ();                                                          // Call test function for decimal to hexadecimal conversion
+   char choice;
+   do {
+      int num = 0, validInput = 0;                                         // Variable to store the user input
+      char buffer[100];
+      while (!validInput) {                                                // Prompt the user for input
+         printf ("Enter a decimal number: ");
+         if (fgets (buffer, sizeof (buffer), stdin)) {
+            char* endptr;
+            num = strtol (buffer, &endptr, 10);
+            if (*endptr == '\0' || *endptr == '\n') validInput = 1;
+            else printf ("Invalid input. Please enter a valid decimal number.\n");
+         }
+         if (!validInput) {
+            printf ("Do you want to try again? (y/n): ");
+            choice = getchar ();
+            while (getchar () != '\n');
+            if (choice != 'y' && choice != 'Y') {
+               printf ("Exiting the program.\n");
+               return 0;
+            }
+         }
+      }
+      char* binaryStr = Dec2Bin (num);                           // Convert the decimal number to binary
+      if (binaryStr != NULL) {
+         printf ("Binary: %s\n", binaryStr);
+         free (binaryStr);
+      }
+      else printf ("Memory allocation failed.\n");
+      char* hexStr = Dec2Hex (num);                                // Convert the decimal number to hexadecimal
+      if (hexStr != NULL) {
+         printf ("Hexadecimal: %s\n", hexStr);
+         free (hexStr);
+      }
+      else printf ("Memory allocation failed.\n");
+      printf ("Decimal: %d\n", num);                                   // Print the decimal number
+      printf ("\nDo you want to enter another number? (y/n): ");       // Ask if user wants to continue
+      choice = getchar ();
+      while (getchar () != '\n');

```

Comment:
You can use fgets to read the full line at once. This way you don't have to have a loop to empty the buffer and you can check for specific inputs.